File: typescript/src/database/InMemoryUserRepository.ts
import { RegisteredUser, UserToRegister } from "../domain/entities/User";
import UserRepository from "../domain/repositories/UserRepository";

export default class InMemoryUserRepository implements UserRepository {
  users: UserToRegister[] = []

  store(user: UserToRegister): void {
    this.users.push(user)
  }

  isUsernameAlreadyUsed(usernameToFind: string): boolean {
    return this.users.some((u: UserToRegister) => u.username === usernameToFind)
  }

  getAll(): RegisteredUser[] {
    return [...this.users]
  }

  reset(): void {
    this.users = []
  }
}

File: java/src/main/java/tech/qmates/openchat/database/InMemoryUserRepository.java
package tech.qmates.openchat.database;

import tech.qmates.openchat.domain.entity.User;
import tech.qmates.openchat.domain.repository.UserRepository;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class InMemoryUserRepository implements UserRepository {

    private final List<User> users = new ArrayList<>();

    @Override
    public void store(User user) {
        this.users.add(user);
    }

    @Override
    public boolean isUsernameAlreadyUsed(String username) {
        return users.stream().anyMatch(user -> user.username().equals(username));
    }

    @Override
    public List<User> getAll() {
        return Collections.unmodifiableList(this.users);
    }

    @Override
    public void reset() {
        this.users.clear();
    }
}
